### ISSUE TITLE
New session: imbalanced data with the poker hand dataset (FFN)

### ISSUE THREAD

**Issue by yoavram** [yoavram/DataSciPy#15]
## Why

`sets.ipynb` in the transformers workshop trains three architectures on the UCI poker hand dataset and hits a result that is really a *data science* lesson rather than an architecture one: **every model scores 0% recall on flush**, while reporting 99%+ accuracy.

That notebook now diagnoses the failure and deliberately stops there, because its subject is attention. It points here for the fix. This issue carries everything already measured so the session can be built without rediscovering any of it.

Suggested file: `sessions/imbalanced.ipynb`, using the **flattened FFN** only — no transformer, no prerequisite from the other workshop.

## The dataset

`data/poker-hand-training-true.data` + `data/poker-hand-testing.data` (UCI id 158), concatenated to 1,025,010 hands. Each row is 5 × (suit, rank) plus a class in 0–9. Already vendored in this repo's `download_data.py` pattern; the legacy URL `archive.ics.uci.edu/ml/machine-learning-databases/poker/...` still returns 200 (verified 2026-09-21).

It is extremely skewed. Training-split counts (80/10/10 split, seed 8765):

| class | train | val |
|---|---:|---:|
| Nothing | 410,972 | 51,427 |
| One pair | 346,373 | 43,432 |
| Two pairs | 39,117 | 4,764 |
| Three of a kind | 17,305 | 2,133 |
| Straight | 3,175 | 394 |
| Flush | 1,663 | 180 |
| Full house | 1,183 | 150 |
| Four of a kind | 198 | 20 |
| Straight flush | 14 | 1 |
| Royal flush | 8 | 0 |

Note the last two rows: **14 and 8 training examples**, and 1 and 0 in validation. They constrain what any result can honestly claim.

## Finding 1 — the ceiling is computable from the data, before any model exists

Group hands by *rank signature* alone (multiset pattern + whether the ranks form a run). This separates 7 of the 10 classes perfectly and leaves exactly two ambiguous groups:

```
(1,1,1,1,1) not a run  ->  Nothing (51,427)  or  Flush (180)
(1,1,1,1,1) a run      ->  Straight (394)    or  Straight flush (1)
```

Both are resolvable only by looking at suits. A flush is *definitionally* a Nothing hand whose five cards happen to share a suit.

So the best possible model that never reads suits scores **99.82% plain accuracy and 77.78% macro (= 7/9)**. Reading suits at all is worth ~0.2% of accuracy — which is why nothing learns it.

```python
from collections import Counter

def rank_signature(hand_ranks):
    pattern = tuple(sorted(Counter(hand_ranks).values(), reverse=True))
    s = sorted(hand_ranks)
    run = bool(len(set(s)) == 5 and (s[4] - s[0] == 4 or s == [0, 9, 10, 11, 12]))
    return pattern, run   # ranks 0-indexed: 0 = Ace

groups = {}
for r, y in zip(val_ranks, val_labels):
    groups.setdefault(rank_signature(r), Counter())[int(y)] += 1
oracle = {sig: cnt.most_common(1)[0][0] for sig, cnt in groups.items()}
```

This is the strongest single idea to carry over: **work out what your task actually requires before attributing a failure to your model.** It costs three lines and no GPU, and in the transformers notebook it predicted the trained model's score to two decimal places.

## Finding 2 — measured FFN results across sampling regimes

Flattened FFN, 20,150 parameters, 50,000 steps, batch 256, `optax.adamw(1e-3, weight_decay=1e-4)`. Only the **minibatch sampler** differs. Validation recall:

| class | support | natural | balanced | sqrt-balanced |
|---|---:|---:|---:|---:|
| Nothing | 51,427 | 99.9% | 95.8% | 99.1% |
| One pair | 43,432 | 99.2% | 89.5% | 99.4% |
| Two pairs | 4,764 | 97.3% | 95.7% | 97.9% |
| Three of a kind | 2,133 | 93.5% | 94.8% | 99.5% |
| Straight | 394 | 100.0% | 100.0% | 99.2% |
| **Flush** | 180 | **0.0%** | **100.0%** | **78.9%** |
| Full house | 150 | 48.7% | 83.3% | 94.7% |
| Four of a kind | 20 | 75.0% | **60.0%** | 95.0% |
| Straight flush | 1 | 0.0% | 0.0% | 0.0% |
| **plain / macro** | | **99.11 / 68.18** | 93.09 / **79.90** | **99.14 / 84.86** |

Four things here are worth a section each:

1. **Plain accuracy is useless on this data.** A majority-class baseline scores ~50%. The natural-sampling FFN looks excellent at 99.11% while scoring zero on a whole class.
2. **The trade-off is real** — full balancing costs 6 points of plain accuracy (99.11 → 93.09) to buy macro.
3. **Aggressive balancing can make things worse.** Four of a kind *drops* from 75.0% to 60.0% under uniform-over-classes. Oversampling a 198-example class into 10% of every batch hurt it.
4. **The gentler intervention dominates.** `sqrt`-balanced (weight ∝ √count) gets the best macro **and** leaves plain accuracy unchanged (99.11 → 99.14). Both metrics improve at once. That is a genuinely non-obvious result and a good thing for students to discover rather than be told.

Sampler used (a padded index pool, jit-friendly):

```python
cls_idx = [np.where(train_y == c)[0] for c in range(10)]
counts  = np.array([len(i) for i in cls_idx])
L = int(counts.max())
pool = np.stack([ix[np.arange(L) % len(ix)] if len(ix) else np.zeros(L, int) for ix in cls_idx])

w = np.sqrt(counts); w = w / w.sum()          # sqrt-balanced; omit for uniform-over-classes
c = jax.random.choice(k1, jnp.arange(10), (B,), p=jnp.array(w))
j = jax.random.randint(k2, (B,), 0, L)
idx = jnp.asarray(pool)[c, j]
```

## Finding 3 — a linear probe separates "can't" from "won't"

Freeze the trained network, take the representation fed to its classifier head, and fit a linear classifier on class-balanced draws. On the Set Transformer this recovered **66.1% flush recall** and a macro of **83.76%**, above the 77.78% suit-blind ceiling — from frozen features, with no retraining. The information was there; the head had no reason to use it.

It is not free: *Nothing* recall collapses to 34.6% and plain accuracy to 66.97%. That is the clean illustration of the trade — a probe stuck with weak features can only move its threshold. Retraining properly (above) shows the trade shrinking or vanishing, because the model can then build features that make "all five share a suit" an exact test rather than a fuzzy one.

Probing is cheap and transfers to any classifier, so it may deserve its own short section.

## Finding 4 — weight decay shrinks channels nothing rewards

RMS weight per entry, both tables initialised at 0.02:

| model | suit_emb | rank_emb | ratio |
|---|---|---|---|
| Set Transformer | 0.0112 (**below** init) | 0.1735 | 15.5× |
| Deep Sets | 0.0289 | 0.5588 | 19.3× |
| Flattened FFN | 0.2262 (above init) | 0.6553 | 2.9× |

Interesting for this session: the **FFN keeps a large suit channel and still never predicts a flush.** So a shrunken embedding is one route to ignoring a feature, not the only one, and weight magnitude is evidence about what training rewarded, not proof of what a model can do. Worth stating explicitly if the notebook shows these numbers.

## Suggested spine

1. Load the data, show the class distribution, train the FFN as-is. It reports 99.11%.
2. Ask what that number hides → per-class recall, macro vs micro, majority baseline. Flush is 0%.
3. Before blaming the model: compute the suit-blind ceiling from the data (Finding 1).
4. Fix it. Compare natural / balanced / √-balanced (Finding 2), reporting plain **and** macro each time. Land the point that the gentle version wins on both, and that the aggressive one hurt a class.
5. Optionally: class-weighted loss and threshold tuning as alternatives that do not resample.
6. Optionally: the linear probe (Finding 3).
7. Close on which metric to report, and to whom.

## Caveats the notebook should state rather than paper over

- **Straight flush (14 train / 1 val) and royal flush (8 train / 0 val) cannot be honestly learned or evaluated.** Under uniform-over-classes each appears in ~10% of every batch, which is memorisation; validation has one and zero examples respectively, so it cannot detect the failure. The Set Transformer reached "100% macro" partly on this basis — impressive-looking and not falsifiable. Recommend reporting macro over the classes with meaningful support and saying why.
- Macro over 9 classes (royal flush absent from validation) is why a majority baseline scores 1/9 ≈ 11.1%, not 10%.
- Deep Sets on this dataset is not reproducible run to run (92.1 / 84.4 / 85.1 / 77.6 across four runs). Irrelevant if the session is FFN-only, but do not import numbers for it.

## Repo fit and cost

- `sessions/FFN.ipynb` is pure JAX with `keras` only for datasets, `sklearn.metrics` for scoring, seaborn for styling. All of the above is JAX + optax, so it transfers directly. `sklearn.metrics.classification_report` and `balanced_accuracy_score` would replace the hand-rolled per-class helpers.
- House style in `sessions/` binds `import jax.numpy as np`. The code above assumes `jnp`. Flagging rather than changing — worth a deliberate decision either way, since the alias has caused confusion elsewhere.
- **Cost: ~232 s per FFN training on CPU** (48 threads). The full comparison above is three trainings, about 12 minutes, so this session can afford several regimes — which the transformers notebook could not, at ~20 minutes per model. That is the main reason this lesson belongs here.

Raw results and the scripts that produced them can be regenerated from the numbers above; ask if the JSON would help.
